### DISCUSSION TITLE
Full editor UI overhaul to make almost everything an independent tab or window

### DISCUSSION THREAD

**Discussion by tektrip-biggles** [godotengine/godot-proposals#9034] [Editor]
I've seen a few discussion of Godot's UI that touch on similar problems but not one that specifically focuses on this. For background, I've not used Unity before, but do have a few years of professional full-time experience with UE4. I also predominantly use a Mac, which may be influencing my thoughts here.

## Main problems 

1. Godot's UI is often extremely inconsistent & can be confusing
2. There are currently 3 "kinds" of main tabs which determine what's shown in the singleton-like "main window": 
  * `2D` / `3D` / `Script` / `AssetLib` horizontal super-mode-switcher at the very top
  * Open "Scene" tabs, always visible & arranged horizontally. 
  * Open "Script" tabs, arranged vertically, but only when in "Script" mode
4. Furthermore, there are other tabs, most notably:
  * The "bottom panel" has tabs for e.g. `Output`, `Debugger`, `Search Results`, `Audio`, `Animation` & `Shader Editor`.
  * Other tabs that can be "docked" to one of the upper or lower side panels include: `Scene`, `FileSystem`, `Import`, `History`, `Inspector` & `Node`. These can be moved & docked elsewhere but only as side-panels. Note that several of these are context-dependent. E.g. the `Inspector` tab can show details for a file selected in the `FileExplorer` tab _or_ a node selected in the `Scene` tab. Others, such as `Import` or `Node` simply don't make sense unless something of the relevant type is selected in one of the other tabs. 
  * The Shader Editor bottom tab also has _its own_ (vertical) list of open shaders
  * The Debugger bottom tab again has its own set of (horizontal) tabs: Stack Trace, Errors, Profiler, Visual Profiler, Monitors, Video RAM, Misc & Network Profiler
5. As a result of all this, it's not possible to:
  * Actually _use_ the "Scene" tabs at the top to navigate between more than a very small handful of open scenes. This can then discourage breaking scenes up into smaller components as it can quickly become very unwieldy to switch between them
  * View two scenes or scripts simultaneously (although by "undocking" the script editor it is at least now possible to view a 2D or 3D scene and a script at the same time)
  * View the console `Output` at the same time as a `Debug` tab. Note that since there are nested Debug tabs, this will often require two clicks to return to the debug view that you were previously looking at.  
  * "Inspect" more than one node and / or file at a time side-by-side
7. Also as a result of all these different tab types being so singleton-like, selecting something in _one_ tab will often prompt the editor to switch tabs in _another_ part of the UI in an effort to try and show the user what it thinks is the most relevant information 
8. The user needs to remember different keyboard shortcuts for switching between different "kinds" of tabs, e.g. switching scenes vs switching scripts. (Note: this isn't helped by the current issue on MacOS where Godot doesn't recognise the OS standard "Ctrl + Tab" & "Ctrl + Shift + Tab" as eligible keyboard shortcuts available for binding to, say, next / previous script respectively)
9. It can also often be unintuitive and difficult to know which _part_ of the UI has focus at any given moment & will respond to key presses or undo / redo etc.
10. A relatively minor one but at first glance it's not always obvious _which script or script file_ is actually being edited in the script editor at any given moment! (This is shown under the list of open scripts in the script editor's side panel though.
11. Once you have more than a trivial number of scenes and scripts open it can be very cumbersome to jump between them. I find myself scrolling constantly through the file explorer window instead of using KB shortcuts or either of the dedicated horizontal & vertical tab bars for these things.

(note that where I refer to a "tab" in the UI, it may be referring to the clickable button and / or the individual "view" that appears when that button is clicked)

## Proposed solution

This is a high-level proposal that I think would address most of these issues but I suspect may not be the easiest to implement. You can probably see where I'm going with this but I think what the UI really needs is to step away from everything being quite so "modal" or "singleton-like". 

Instead I'd propose there be a _single unified tab bar_ running the full width of (each) Godot window which can contain any number of different kinds of tabs. With a few exceptions I think that pretty much _every_ one of the different kinds of "tabs" mentioned in the previous section could be treated like one of these top-level web browser tab. 

This would be supported by common standard modern UI navigation features such as:
1. Ability to "undock" any tab into its own OS-level Godot window, or dock it _into_ any already existing OS-level Godot window.
2. Ability to open a selected file (or scene or script or node etc) in the "current" tab or in a "new tab" via right mouse menu or "Ctrl + Click" or similar
3. Ability to quickly switch between tabs in the active window using either "tab order" or "history order" (configured by user preference) with "Ctrl + Tab" & "Ctrl + Shift + Tab" or similar. (Note: since there may well be a lot of horizontal tabs in any given window it may be good for convenience to show these in a "vertical list" after pressing one of these shortcuts but before releasing the "Ctrl" button)
4. Ability to arbitrarily "split" an OS-level window into partitions, each with their own "horizontal tab bar". I would expect the "default" window layout to make use of this in order to partially replicate the existing _layout_ (if not behaviour) of the editor UI
5. Perhaps the ability to toggle any given tab bar between "horizontal" and "vertical" modes.
6. A clear visual indication as to which window has the active UI focus at any given moment

Some tabs it may still only ever make sense to ever have one of (e.g. `Output`) and there are some tabs that users may not care about. UE4's solution to this is to have a top-level "Windows" menu item with a list of possible singleton-like windows which allows the user to go straight to any of these & also shows tick marks next to those that are already open somewhere. Some are grouped under sub-menus in order to keep things nicely organised & it even provides up to 4 different (singleton-like) "Find in Blueprints" windows so that you can keep the results of 3 previous searches open while doing a fourth.

Also, some of the current tabs do genuinely only make sense as a part of a particular "mode" and so probably _wouldn't_ make sense as being top-level tabs. I'm thinking things like the `Scene` (scene tree), `Inspector` & `Node` tabs which should probably each be components of other top-level tabs. I.e. _each_ 2D or 3D scene's tab would have its own `Scene`, `Inspector` & `Node` panel. So if you had two scenes open at the same time, you could select a node in each & inspect both simultaneously. Switching to a tab with a script on the other hand wouldn't show any of these by default.

Note that this proposal is similar to how both UE4 & Blender do their editor window management & I'm not necessarily proposing copying either in full, but I think there are good arguments for moving in this direction & I'm keen to hear what people think, whether for or against something like this.

Of course there may be good reasons for the current UI window management philosophy that I'm unaware of but it's by far been my biggest frustration with using Godot so far.

**Comment by Mortasen** [comment]
I’m honestly surprised this proposal hasn’t gotten more attention - both from the community and the maintainers. The problems with the current UI were brought up in countless issues here and on other platforms, and this proposal has the potential to actually solve pretty much all of them in a very meaningful way.

The idea of having a more unified, modular tab system just makes so much sense. The current setup, with its different types of tabs and panels that don’t always interact intuitively, feels restrictive and clunky. Being able to open scenes, scripts, and even debugging tools side-by-side without constantly flipping back and forth would be such a game-changer for workflow. I can imagine how much easier it would be to juggle multiple tasks - debugging, editing a script, tweaking a node - without feeling like you’re fighting the editor itself.

That said, I completely understand if the maintainers don’t have the time to take on something as big as this right now. A UI overhaul of this scale is no small task. But even if it’s not something that can be prioritized officially, it would be great to at least get an opinion on the idea. If it’s something the maintainers aren’t opposed to, maybe someone else in the community would be inspired to pick it up and start working on it.

**Reply by Calinou** [reply]
There are already several open proposals on the subject:

- https://github.com/godotengine/godot-proposals/issues/1508
- https://github.com/godotengine/godot-proposals/issues/7233
- https://github.com/godotengine/godot-proposals/issues/8264

However, [Godot's design favors dedicated solutions as opposed to providing the absolute highest flexibility](https://docs.godotengine.org/en/stable/contributing/development/best_practices_for_engine_contributors.html#to-each-problem-its-own-solution) (which can be a detriment for usability). This is why proposals like https://github.com/godotengine/godot-proposals/issues/4565 are more favored among maintainers overall.

In any case, UI rework proposals are often controversial, which can explain the slow progress in this area.

PS: Remember that you can use multiple Godot instances on the same project at once. This can be used to edit multiple scenes in parallel, but be careful about file conflicts when saving.

**Reply by tektrip-biggles** [reply]
Yeah, I understand that Godot has a different philosophy to this sort of thing & Juan's proposal (from 3 years ago) that you linked to essentially says several times that the UI "works well" & there's little hope of change if people don't even see it as a problem.

And maybe for many users it really isn't a problem so more power to them. For my part, I've gone back to Unreal which I feel far more at home with. Until this discussion topic got resurrected I'd almost forgotten how frustrating my year or two trying to use Godot had been!

**Comment by Shadowblitz16** [comment]
Godot's editor ui is not the thing that needs an overhaul it's godot's control nodes and theming system that do.
